src/core/server/plugins/discovery/plugins_discovery.test.ts

```diff
@@ -17,251 +17,376 @@
+
+describe('plugins discovery system', () => {
+  let logger: ReturnType<typeof loggingServiceMock.create>;
+  let env: Env;
+  let configService: ConfigService;
+  let pluginConfig: PluginsConfigType;
+  let coreContext: CoreContext;
+
+  beforeEach(async () => {
+    logger = loggingServiceMock.create();
+
+    mockPackage.raw = packageMock;
+
+    env = Env.createDefault(
+      getEnvOptions({
+        cliArgs: { envName: 'development' },
+      })
+    );
+
+    configService = new ConfigService(
+      rawConfigServiceMock.create({ rawConfig: { plugins: { paths: [] } } }),
+      env,
+      logger
+    );
+    await configService.setSchema(config.path, config.schema);
+
+    coreContext = {
+      coreId: Symbol(),
+      configService,
+      env,
+      logger,
+    };
+
+    pluginConfig = await configService
+      .atPath<PluginsConfigType>('plugins')
+      .pipe(first())
+      .toPromise();
+
+    // jest relies on the filesystem to get sourcemaps when using console.log
+    // which breaks with the mocked FS, see https://github.com/tschaub/mock-fs/issues/234
+    // hijacking logging to process.stdout as a workaround for this suite.
+    jest.spyOn(console, 'log').mockImplementation((...args) => {
+      process.stdout.write(args + '\n');
+    });
   });
 
-  mockStat.mockImplementation((path, cb) => {
-    if (path.includes('9-inaccessible-dir')) {
-      cb(new Error(`ENOENT (disappeared between "readdir" and "stat").`));
-    } else {
-      cb(null, { isDirectory: () => !path.includes('non-dir') });
-    }
+  afterEach(() => {
+    mockFs.restore();
+    // restore the console.log behavior
+    jest.restoreAllMocks();
   });
 
-  mockReadFile.mockImplementation((path, cb) => {
-    if (path.includes('no-manifest')) {
-      cb(new Error('ENOENT'));
-    } else if (path.includes('invalid-manifest')) {
-      cb(null, Buffer.from('not-json'));
-    } else if (path.includes('incomplete-manifest')) {
-      cb(null, Buffer.from(JSON.stringify({ version: '1' })));
-    } else if (path.includes('incompatible-manifest')) {
-      cb(null, Buffer.from(JSON.stringify({ id: 'plugin', version: '1' })));
-    } else {
-      cb(
-        null,
-        Buffer.from(
-          JSON.stringify({
-            id: 'plugin',
-            configPath: ['core', 'config'],
-            version: '1',
-            kibanaVersion: '1.2.3',
-            requiredPlugins: ['a', 'b'],
-            optionalPlugins: ['c', 'd'],
-            server: true,
-          })
-        )
-      );
-    }
+  it('discovers plugins in the search locations', async () => {
+    const { plugin$ } = discover(new PluginsConfig(pluginConfig, env), coreContext);
+
+    mockFs(
+      {
+        [KIBANA_ROOT]: {
+          src: {
+            plugins: {
+              plugin_a: Plugins.valid('pluginA'),
+            },
+          },
+          plugins: {
+            plugin_b: Plugins.valid('pluginB'),
+          },
```

Comment:
IMO it's a bit easier to grok this if we used regular directory paths as keys rather than nested objects (which mock-fs supports):

```ts
mockFs({
  [KIBANA_ROOT]: {
    'src/plugins/plugin_a': Plugins.valid('pluginA'),
    'plugins/plugin_b': /** */,
     'x-pack/plugins/plugin_c': /** */,
  }
})
```

Reply:
It's supposed to, however I'm having failures when running your changes.

When looking at the actual implementation of `mock-fs`'s filesystem, it seems the `populate` logic does not accept nested multi-path keys (...)

This works:

```ts
    mockFs(
      {
        [`${KIBANA_ROOT}/src/plugins/plugin_a`]: Plugins.valid('pluginA'),
        [`${KIBANA_ROOT}/plugins/plugin_b`]: Plugins.valid('pluginB'),
      },
      { createCwd: false }
    );
```

but this doesn't:

```ts
mockFs(
      {
        [KIBANA_ROOT]: {
          'src/plugins/plugin_a': Plugins.valid('pluginA'),
          'plugins/plugin_b': Plugins.valid('pluginB'),
        },
      },
      { createCwd: false }
    );
```

wdyt? Do we still prefer the long

```ts
[`${KIBANA_ROOT}/src/plugins/plugin_a`]: Plugins.valid('pluginA'),
```

syntax over the one currently used?

Reply:
I think that would be slight more clear, yes.